state/presence/presence.go

```diff
@@ -757,6 +861,174 @@ func (p *Pinger) ping() (err error) {
+}
+
+// flushRemovals makes sure that we've applied all desired removals
+func (p *BeingPruner) flushRemovals() error {
+	if len(p.toRemove) == 0 {
+		return nil
+	}
+	matched, err := p.beingsC.RemoveAll(bson.M{"_id": bson.M{"$in": p.toRemove}})
+	p.toRemove = p.toRemove[:0]
+	if matched.Removed > 0 {
+		p.removedCount += uint64(matched.Removed)
+	}
+	return err
+}
+
+func (p *BeingPruner) removeOldPings() error {
+	// now and now-period are both considered active slots, so we don't
+	// touch those. We also leave 2 more slots around
+	startTime := time.Now()
+	// TODO(jam): 2017-04-18 startCount may not be appropriate, because
+	// Count() includes all models, not just the one we are pruning
+	startCount, err := p.pingsC.Count()
+	if err != nil {
+		return err
+	}
+	logger.Tracef("pruning %q for %q starting with %d docs",
+		p.pingsC.Name, p.modelUUID, startCount)
+	s := timeSlot(time.Now(), p.delta)
+	oldSlot := s - 3*period
+	res, err := p.pingsC.RemoveAll(bson.D{{"_id", bson.RegEx{"^" + p.modelUUID, ""}},
+					     {"slot", bson.M{"$lt": oldSlot}} })
+	if err != nil && err != mgo.ErrNotFound {
+		logger.Errorf("error removing old entries from %q: %v", p.pingsC.Name, err)
+		return err
+	}
+	endCount, _ := p.pingsC.Count()
+	logger.Debugf("pruned %q for %q (with %d docs) of %d old pings down to %d docs in %v",
+		p.pingsC.Name, p.modelUUID, startCount, res.Removed, endCount, time.Since(startTime))
+	return nil
+}
+
+func (p *BeingPruner) removeUnusedBeings() error {
+	var keyInfo collapsedBeingsInfo
+	// TODO(jam): 2017-04-18 startCount may not be appropriate, because
+	// Count() includes all models, not just the one we are pruning
+	startCount, err := p.beingsC.Count()
+	if err != nil {
+		return err
+	}
+	logger.Tracef("pruning %q for %q starting with %d docs",
+		p.beingsC.Name, p.modelUUID, startCount)
+	startTime := time.Now()
+	keyCount := 0
+	seqCount := 0
+	iter := p.iterKeys()
+	for iter.Next(&keyInfo) {
+		keyCount += 1
+		// Find the max
+		maxSeq := int64(-1)
+		for _, seq := range keyInfo.Seqs {
+			if seq > maxSeq {
+				maxSeq = seq
+			}
+		}
+		// Queue everything < max to be deleted
+		for _, seq := range keyInfo.Seqs {
+			seqCount++
+			if seq >= maxSeq {
+				// It shouldn't be possible to be > at this point
+				continue
+			}
+			p.queueRemoval(seq)
+			if len(p.toRemove) > p.maxQueue {
+				if err := p.flushRemovals(); err != nil {
+					return err
+				}
+			}
+		}
+	}
+	if err := p.flushRemovals(); err != nil {
+		return err
+	}
+	if err := iter.Close(); err != nil {
+		return err
+	}
+	logger.Debugf("pruned %q for %q (with %d docs) of %d sequence keys (evaluated %d) from %d keys in %v",
+		p.beingsC.Name, p.modelUUID, startCount, p.removedCount, seqCount, keyCount, time.Since(startTime))
+	return nil
+}
+
+// Prune removes beings from the beings collection that have been superseded by
+// another entry with a higher sequence. It does *not* attempt to remove beings that
+// have not been referenced by recent pings.
```

Comment:
We don't try to avoid recent references (with the premise that it doesn't matter, because the new sequence takes precedence anyway). That was the "it does *not*" section, but obviously I wasn't very clear with my comment.

Would you be more comfortable if we preserve all referenced sequences? If we prune pings first, we can read the remaining pings and build up all the sequence indicators that are still left.
It will be more expensive, because we are transforming the data (from bit-packed fields to a bunch of integers).